File: 剑指offer/4.二维数组中的查找.py
from typing import List


class Solution:
    def findNumberIn2DArray(self, matrix: List[List[int]],
                            target: int) -> bool:
        n = len(matrix)

        if n == 0:
            return False

        m = len(matrix[0])
        if m == 0:
            return False

        row = 0
        column = m - 1

        while row < n and column >= 0:
            if matrix[row][column] == target:
                return True
            elif matrix[row][column] > target:
                column -= 1
            else:
                row += 1

        return False

File: 剑指offer/4.二维数组中的查找.cpp
#include <vector>
using namespace std;

class Solution {
public:
    bool findNumberIn2DArray(vector<vector<int>>& matrix, int target)
    {
        int m = matrix.size();
        if (m == 0) {
            return false;
        }

        int n = matrix[0].size();
        if (n == 0) {
            return false;
        }

        int hang = m - 1;
        int lie = 0;

        while (hang >= 0 && lie < n) {
            if (matrix[hang][lie] == target) {
                return true;
            } else if (matrix[hang][lie] > target) {
                hang -= 1;
            } else if (matrix[hang][lie] < target) {
                lie += 1;
            }
        }
        return false;
    }
};